src/Frame.h

```diff
@@ -82,6 +90,65 @@ class Frame : public BroObj {
 	bool delayed;
 };
 
+/*
+Class that allows for lookups in both a closure frame and a regular frame
+according to a list of IDs passed into the constructor.
+
+Facts:
+	- A ClosureFrame is created from two frames: a closure and a regular frame.
+	- ALL operations except GetElement operations operate on the regular frame.
+	- A ClosureFrame requires a list of outside ID's captured by the closure.
+	- Get operations on those IDs will be performed on the closure frame.
+
+ClosureFrame allows functions that generate functions to be passed between
+different sized frames and still properly capture their closures. It also allows for
+cleaner handling of closures.
+*/
+class ClosureFrame : public Frame {
+public:
+	ClosureFrame(Frame* closure, Frame* body,
+		std::shared_ptr<id_list> outer_ids);
+	~ClosureFrame() override;
+	Val* GetElement(ID* id) const override;
+	void SetElement(const ID* id, Val* v) override;
+	Frame* Clone() override;
+        Frame* SelectiveClone(id_list* selection) override;
+
+private:
+	Frame* closure;
+	Frame* body;
+
+	// Searches this frame and all sub-frame's closures for a value corresponding
+	// to the id.
+        static Val* GatherFromClosure(const Frame* start, const ID* id);
+        // Moves through the closure frames and associates val with id.
+        static void SetInClosure(Frame* start, const ID* id, Val* val);
+
+        bool closure_contains(const ID* i) const
+            { return this->closure_elements.find(i->Name()) != this->closure_elements.end(); }
+
+        // Hashes c style strings. The strings need to be null-terminated.
+	struct const_char_hasher {
+		size_t operator()(const char* in) const
+			{
+			// http://www.cse.yorku.ca/~oz/hash.html
+			size_t h = 5381;
+			int c;
+
+			while ((c = *in++))
+				h = ((h << 5) + h) + c;
+
+			return h;
+			}
+	};
+
+	// NOTE: In a perfect world this would be best done with a trie or bloom
+	// filter. We only need to check if things are NOT in the closure.
+        // In reality though the size of a closure is small enough that operatons are
+        // fairly quick anyway.
+	std::unordered_set<const char*, const_char_hasher> closure_elements;
```

Comment:
In fact the set will typically be tiny, meaning a vector with linear search might be the most efficient thing.

Reply:
Good point :) I updated that.